src/elysium/test/sigmawallet_tests.cpp

```diff
@@ -102,6 +109,98 @@ class TestSigmaWallet : public SigmaWallet
+    {
+        SigmaPrivateKey priv;
+
+        // first 32 bytes as seed
+        uint256 serialSeed;
+        std::copy(seed.begin(), seed.begin() + 32, serialSeed.begin());
+        priv.serial.memberFromSeed(serialSeed.begin());
+
+        // last 32 bytes as seed
+        uint256 randomnessSeed;
+        std::copy(seed.begin() + 32, seed.end(), randomnessSeed.begin());
+        priv.randomness.memberFromSeed(randomnessSeed.begin());
+
+        return priv;
+    }
+
+    class Database : public SigmaWallet::Database {
+    public:
+        Database()
+        {
+        }
+
+    public:
+        bool WriteMint(SigmaMintId const &id, SigmaMint const &mint, CWalletDB *db = nullptr)
+        {
+            auto local = Connection(db);
+            return local->WriteElysiumMintV0(id, mint);
+        }
+
+        bool ReadMint(SigmaMintId const &id, SigmaMint &mint, CWalletDB *db = nullptr) const
+        {
+            auto local = Connection(db);
+            return local->ReadElysiumMintV0(id, mint);
+        }
+
+        bool EraseMint(SigmaMintId const &id, CWalletDB *db = nullptr)
+        {
+            auto local = Connection(db);
+            return local->EraseElysiumMintV0(id);
+        }
+
+        bool HasMint(SigmaMintId const &id, CWalletDB *db = nullptr) const
+        {
+            auto local = Connection(db);
+            return local->HasElysiumMintV0(id);
+        }
+
+        bool WriteMintId(uint160 const &hash, SigmaMintId const &mintId, CWalletDB *db = nullptr)
+        {
+            auto local = Connection(db);
+            return local->WriteElysiumMintIdV0(hash, mintId);
+        }
+
+        bool ReadMintId(uint160 const &hash, SigmaMintId &mintId, CWalletDB *db = nullptr) const
+        {
+            auto local = Connection(db);
+            return local->ReadElysiumMintIdV0(hash, mintId);
+        }
+
+        bool EraseMintId(uint160 const &hash, CWalletDB *db = nullptr)
+        {
+            auto local = Connection(db);
+            return local->EraseElysiumMintIdV0(hash);
+        }
+
+        bool HasMintId(uint160 const &hash, CWalletDB *db = nullptr) const
+        {
+            auto local = Connection(db);
+            return local->HasElysiumMintIdV0(hash);
+        }
+
+        bool WriteMintPool(std::vector<MintPoolEntry> const &mints, CWalletDB *db = nullptr)
+        {
+            auto local = Connection(db);
+            return local->WriteElysiumMintPoolV0(mints);
+        }
+
+        bool ReadMintPool(std::vector<MintPoolEntry> &mints, CWalletDB *db = nullptr)
+        {
+            auto local = Connection(db);
+            return local->ReadElysiumMintPoolV0(mints);
+        }
+
+        void ListMints(std::function<void(SigmaMintId&, SigmaMint&)> const &inserter, CWalletDB *db = nullptr)
+        {
+            auto local = Connection(db);
+            local->ListElysiumMintsV0<SigmaMintId, SigmaMint>(inserter);
+        }
+    };
```

Comment:
why are we copying above logic here?

Reply:
The purpose of this is testing the abstract parent class of wallet v0 and v1, you can think that logic above is mocking functions to test the main functionality. Actually, the better way to do is defining separated functions instead of relied on DB functions of v0 but we have a lot of things to do if we need it. Anyway, I think we could keep it like this and add a todo comment to remove DB dependency when we have to change the DB functions of v0.

Reply:
The purpose of this is to test the abstract parent class of wallet v0 and v1, then we need to define all undefined functions of the class. Most logics are the same with v0 but the reason that I don't test through v0 because I need v0 to be able to change without worrying about the parent class and the tests.